File: comms/native/services/networks/PinTraceNode.js
// Copyright: IQGeo Limited 2010-2023

//import {MywError} from 'myWorld-native-services'
import ProgressHandler from '../base/ProgressHandler';
import { TraceNode } from 'myWorld-native-services';
import TracePin from './TracePin';
import geomUtils from '../base/GeomUtils';
import myw from 'myWorld-base';

/*eslint-disable no-await-in-loop*/
class PinTraceNode extends TraceNode {
    static {
        /**
         * Superclass for nodes in a pin-level network trace result
         *
         * Provides methods for trace consolidation
         */

        // User-level properties (for all subclasses)
        this.prototype.metadata = [
            'ports',
            'fibers',
            'from_',
            'to_',
            'length',
            'direction',
            'individualLoss',
            'cumulativeLoss',
            'loc',
            ...TraceNode.prototype.metadata
        ];

        this.prototype.metadata_unit_scales = {
            ...TraceNode.prototype.metadata_unit_scales,
            length: { scale: 'length', unit: 'm' },
            individualLoss: { scale: 'fiber_loss', unit: 'dB' },
            cumulativeLoss: { scale: 'fiber_loss', unit: 'dB' }
        };

        // Backstop values for user-level properties
        this.prototype.ports = undefined;

        this.prototype.fibers = undefined;
        this.prototype.from_ = undefined;
        this.prototype.to_ = undefined;
        this.prototype.loc = undefined;

        // For debugging
        this.prototype.progress = ProgressHandler.newFor('comms.tracing', 0);
    }

    /**
     * Init slots of self
     *
     * PINS in a PinRange. DIRECTION is trace direction ('upstream', 'downstream' or 'both')
     */
    constructor(feature, tech, pins, direction, dist, parent = undefined) {
        super(feature, dist, /*parent:*/ parent);
        this.pins = pins;
        this.direction = direction;
        this.tech = tech;
    }

    //@property
    /**
     * Length of self's leg, in metres (if appropriate)
     */
    _length() {
        let length;
        if (this.parent) {
            length = this.dist - this.parent.dist;
        } else {
            length = this.dist;
        }

        if (!length) {
            return undefined;
        }

        return length.toFixed(2);
    }

    /**
     * Individual Loss in dB.
     */
    get individualLoss() {
        const featureFunction = this._functionOf();
        if (!featureFunction) return 0.0;

        const lossConfig = myw.config['mywcom.loss'];
        const lossByTech = lossConfig[this.tech];
        if (!lossByTech) return 0.0;

        let loss = lossByTech[featureFunction];
        if (!loss) return 0.0;

        if (this.tech === 'copper' && featureFunction === 'cable') {
            loss = loss[String(parseInt(this.feature.properties.gauge))] || 0.0;
        }

        return this.length ? this.length * (loss / 1000) : loss;
    }

    /**
     * @property
     * Cumulative Fiber Loss in dB (iterative)
     * @returns  {float}
     */
    get cumulativeLoss() {
        if (this.cumulative_loss) {
            return this.cumulative_loss;
        }

        let node = this;
        let cumulative_loss = 0.0;

        while (node) {
            if (node.parent) {
                // prevent adding duplicative loss to cumulative loss
                if (node.feature.getUrn() !== node.parent.feature.getUrn()) {
                    cumulative_loss += node.individualLoss;
                }
            } else {
                cumulative_loss += node.individualLoss;
            }
            node = node.parent;
        }

        this.cumulative_loss = cumulative_loss;

        return this.cumulative_loss;
    }

    /**
     * The coordinates for path from this' root node to this (iterative)
     * @returns {Array} coords
     */
    coordsFromRoot() {
        const coords = [];

        // add this
        this._addCoords(this, coords);

        let currentNode = this;
        while (currentNode.parent) {
            this._addCoords(currentNode.parent, coords);
            currentNode = currentNode.parent;
        }

        return coords.reverse();
    }

    /**
     * Adds coordinate pairs to coords from currentNode
     * @param {Object} currentNode
     * @param {Array} coords list of coordinates to add to
     */
    _addCoords(currentNode, coords) {
        const nodeCoords = currentNode.coords();

        // case: linestring so reverse it's coords
        if (nodeCoords.length > 1) {
            nodeCoords.reverse();
        }

        for (let coord of nodeCoords) {
            if (!coords.length || !geomUtils.coordEqual(coord, coords[coords.length - 1])) {
                coords.push(coord);
            }
        }
    }

    /**
     * The coordinates of self's path
     *
     * Returns a list of coords
     */
    coords() {
        return this.feature.geometry.flatCoordinates();
    }

    /**
     * Consolidate consecutive links in self's sub-tree
     *
     * Returns self
     */
    async tidy() {
        // Convert cable segs -> cables
        await this.mutateCableSegments();

        // Consolidate nodes (to simplify result)
        const nodes = [this];
        while (nodes.length) {
            const node = nodes.pop();
            node.consolidate();
            for (const child of node.children) {
                nodes.push(child);
            }
        }

        return this;
    }

    /**
     * Convert cable segment nodes of self's subtree to cable nodes
     *
     * Returns self
     */
    // Uses pseudo-recursion to avoid stack overflow
    async mutateCableSegments() {
        const nodes = [this];

        while (nodes.length) {
            const node = nodes.shift();

            if (node.type == 'segment') {
                await node.convertToCable();
            }

            for (const child of node.children) {
                nodes.push(child);
            }
        }

        return this;
    }

    /**
     * Merge self's 'obvious' children into self (to simplify trace)
     */
    // Subclassed in MywcomSegmentTraceNode
    consolidate() {
        // If child is single connection .. skip it
        if (this.children.length == 1) {
            const child = this.children[0];

            if (child.type == 'connection' && child.children.length == 1 && !child.conn.is_splice) {
                this.children = child.children;
            }
        }
    }

    /**
     * The leaf nodes of self's sub-tree (iterative)
     * ENH: move to super
     * @returns {Array}
     */
    leafNodes() {
        const leafNodes = [];
        const stack = [this];

        while (stack.length > 0) {
            const currentNode = stack.pop();

            if (!currentNode.children || currentNode.children.length === 0) {
                leafNodes.push(currentNode);
            } else {
                stack.push(...currentNode.children);
            }
        }

        return leafNodes;
    }

    /**
     * The trace pin at which each of self's pins terminates (recursive)
     *
     * Returns a list of MywcomTracePins, keyed by pin number
     */
    terminations() {
        // ENH: handle multiple terminations (splitting etc)

        const trace_pins = {};

        // Add leaf pins from children
        for (const child_node of this.children) {
            const child_trace_pins = child_node.terminations();

            for (const child_pin in child_trace_pins) {
                const pin = this.pinFor(child_node, child_pin);
                if (!trace_pins[pin]) {
                    trace_pins[pin] = child_trace_pins[child_pin];
                    this.progress(5, this, 'Mapped pin', pin, '->', child_node, child_pin);
                }
            }
        }

        // Add self's leaf pins
        for (const pin of this.pins.range()) {
            if (!trace_pins[pin]) {
                trace_pins[pin] = new TracePin(this, pin);
            }
        }

        return trace_pins;
    }

    /**
     *
     * @returns
     */
    _functionOf() {
        let functionDef = (this.equipment = myw.config['mywcom.equipment'][this.feature.getType()]);

        if (!functionDef && myw.config['mywcom.cables'][this.feature.getType()])
            functionDef = {
                function: 'cable'
            };

        if (
            !functionDef &&
            (this.feature.getType() === 'mywcom_fiber_connection' ||
                this.feature.getType() === 'mywcom_copper_connection') &&
            this.feature.properties.splice
        )
            functionDef = {
                function: 'splice'
            };

        return functionDef?.function;
    }
}

Object.defineProperty(PinTraceNode.prototype, 'length', {
    get() {
        return this._length();
    }
});

export default PinTraceNode;

File: comms/server/networks/pin_trace_node.py
# Copyright: IQGeo Limited 2010-2023

import myworldapp.core.server.controllers.base.myw_globals as myw_globals
from myworldapp.core.server.base.core.myw_error import MywError
from myworldapp.core.server.base.core.myw_progress import MywSimpleProgressHandler
from myworldapp.core.server.networks.myw_trace_node import MywTraceNode
from .trace_pin import TracePin


class PinTraceNode(MywTraceNode):
    """
    Superclass for nodes in a pin-level network trace result

    Provides methods for trace consolidation"""

    # User-level properties (for all subclasses)
    metadata = [
        "ports",
        "fibers",
        "from_",
        "to_",
        "length",
        "direction",
        "individualLoss",
        "cumulativeLoss",
        "loc",
    ] + MywTraceNode.metadata

    metadata_unit_scales = dict(
        MywTraceNode.metadata_unit_scales,
        **{
            "length": {"scale": "length", "unit": "m"},
            "individualLoss": {"scale": "fiber_loss", "unit": "dB"},
            "cumulativeLoss": {"scale": "fiber_loss", "unit": "dB"},
        },
    )

    # Used for tracking cumulative loss
    cumulative_loss = None

    # Backstop values for user-level properties
    ports = None
    fibers = None
    from_ = None
    to_ = None
    loc = None

    # For debugging
    progress = MywSimpleProgressHandler(0)

    def __init__(self, feature, tech, pins, direction, dist, parent=None):
        """
        Init slots of self

        PINS in a PinRange. DIRECTION is trace direction ('upstream', 'downstream' or 'both')"""

        super().__init__(feature, dist, parent=parent)
        self.pins = pins
        self.direction = direction
        self.db = myw_globals.db
        self.tech = tech # tech defined at network engine


    @property
    def length(self):
        """
        Length of self's leg, in metres (if appropriate)
        """

        if self.parent:
            length = self.dist - self.parent.dist
        else:
            length = self.dist

        if not length:
            return None

        return "{:.2f}".format(length)

    @property
    def individualLoss(self):
        """
        Individual loss in dB.
        """
        feature_function = self.function()
        if not feature_function:
            return 0.0

        loss_config = self.db.setting("mywcom.loss")
        loss_by_tech = loss_config.get(self.tech)
        if loss_by_tech == None: 
            return 0.0
        
        loss = loss_by_tech.get(feature_function) 
        if loss == None:
            return 0.0

        # If copper cable, must account for gauge
        if self.tech == "copper" and feature_function == "cable":
            loss = loss.get(str(int(self.feature.gauge)), None)
            if loss == None:
                return 0.0
            
        return loss if not self.length else float(self.length) * (loss / 1000)
    
    @property
    def cumulativeLoss(self):
        """
        Cumulative fiber loss in dB. (iterative)
        """

        if self.cumulative_loss:
            return self.cumulative_loss

        node = self
        cumulative_loss = 0.0

        while node is not None:
            if node.parent:
                # prevent adding duplicative loss to cumulative loss
                if node.feature._urn() != node.parent.feature._urn():
                    cumulative_loss += node.individualLoss
            else:
                cumulative_loss += node.individualLoss
            node = node.parent

        self.cumulative_loss = cumulative_loss

        return self.cumulative_loss

    def coordsFromRoot(self):
        """
        The coordinates for path from self's root node to self (iterative)

        Returns a list of coords
        """
        coords = []

        # add self
        self._addCoords(self, coords)

        current_node = self
        while current_node.parent:
            self._addCoords(current_node.parent, coords)
            current_node = current_node.parent

        return list(reversed(coords))

    def _addCoords(self, current_node, coords):
        """
        Appends coord to COORDS
        """
        node_coords = current_node.coords()

        # case: linestring so reverse it's coords
        if len(node_coords) > 1:
            node_coords.reverse()

        for coord in node_coords:
            if not coords or coord != coords[-1]:
                coords.append(coord)

    def coords(self):
        """
        The coordinates of self's path

        Returns a list of coords"""

        return self.feature._primary_geom_field.geom().coords

    def tidy(self):
        """
        Consolidate consecutive links in self's sub-tree

        Returns self"""

        # Convert cable segs -> cables
        self.mutateCableSegments()

        # Consolidate nodes (to simplify result)
        nodes = [self]
        while nodes:
            node = nodes.pop()
            node.consolidate()
            nodes += node.children

        return self

    def mutateCableSegments(self):
        """
        Convert cable segment nodes of self's subtree to cable nodes

        Returns self"""
        # Uses pseudo-recursion to avoid stack overflow

        nodes = [self]

        while nodes:
            node = nodes.pop(0)

            if node.type == "segment":
                node.convertToCable()

            for child in node.children:
                nodes.append(child)

        return self

    def consolidate(self):
        """
        Merge self's 'obvious' children into self (to simplify trace)
        """
        # Subclassed in SegmentTraceNode

        # If child is single connection .. skip it
        if len(self.children) == 1:
            child = self.children[0]

            if (
                child.type == "connection"
                and len(child.children) == 1
                and not child.conn.is_splice
            ):
                self.children = child.children

    def leafNodes(self):
        """
        The leaf nodes of self's sub-tree (iterative)

        Returns a list of nodes

        ENH: move to super
        """

        leaf_nodes = []
        stack = [self]

        while stack:
            current_node = stack.pop()

            if not current_node.children:
                leaf_nodes.append(current_node)
            else:
                stack.extend(current_node.children)

        return leaf_nodes

    def terminations(self):
        """
        The trace pin at which each of self's pins terminates (non-recursive)
        """

        stack = []
        (current, child_node, children, trace_pins) = (
            self,
            None,
            self.children.copy(),
            {},
        )

        # Reverse list to conform to iterator behaviour; seems going through list in different
        # ways can give different results probably where multiple terminations are present.
        children.reverse()

        while True:

            if not children:
                # Finished with children so process node and add current's pins to tracepins

                # Add leaf pins from current node
                for pin in current.pins.range():
                    if not pin in trace_pins:
                        trace_pins[pin] = TracePin(current, pin)

                if not stack:
                    return trace_pins

                # Pop stack and continue processing parent
                child_trace_pins = trace_pins
                (current, child_node, children, trace_pins) = stack.pop()

                # Add leaf pins from child
                if child_node:
                    for child_pin, trace_pin in child_trace_pins.items():
                        pin = current.pinFor(child_node, child_pin)
                        if not pin in trace_pins:  # ENH: handle multiple terminations
                            trace_pins[pin] = child_trace_pins[child_pin]
                            current.progress(
                                5,
                                current,
                                "Mapped pin",
                                pin,
                                "->",
                                child_node,
                                child_pin,
                            )

            else:
                # Move onto next child
                child_node = children.pop()

                # Push parent and switch to child
                stack.append((current, child_node, children, trace_pins))
                (current, child_node, children, trace_pins) = (
                    child_node,
                    None,
                    child_node.children.copy(),
                    {},
                )
                children.reverse()

    def function(self):
        function_def = self.db.setting("mywcom.equipment").get(
            self.feature.feature_type
        )

        if not function_def and self.db.setting("mywcom.cables").get(
            self.feature.feature_type
        ):
            function_def = {"function": "cable"}

        if (
            not function_def
            and (self.feature.feature_type == "mywcom_fiber_connection" or 
                 self.feature.feature_type == "mywcom_copper_connection")
            and self.feature.splice
        ):
            function_def = {"function": "splice"}

        return function_def["function"] if function_def else None

    def toplevelAStarFeature(self, db_view):
        """
        Return top level feature for purpose of calculating A-star metric
        """

        if hasattr(self.feature, "root_housing"):
            return self.feature.root_housing
        else:
            return self.feature._urn()
